### PR TITLE
Add unique simulation names

### PR DESCRIPTION
**Pull request by collijk**
### Description

*Category*: feature
*JIRA issue*: N/A

This is the first step to fixing a couple issues with simulation logging described here: https://github.com/ihmeuw/vivarium/issues/282 and here: https://github.com/ihmeuw/vivarium/issues/283.
It adds an attribute to th simulation context to track all created contexts within a python process which
allows us to uniquely name simulation contexts.  

This implementation requires that SimulationContexts
are created in serial to ensure thread safety, which I think is a reasonable constraint given all historical 
use cases (in which we never do multiprocessing or multithreading to begin with).  

### Testing
CI + new unit tests

### CODE REVIEW

File: tests/framework/test_engine.py [hunk 1]
```diff
@@ -25,6 +26,12 @@ def is_same_object_method(m1, m2):
     return m1.__func__ is m2.__func__ and m1.__self__ is m2.__self__
 
 
+@pytest.fixture()
+def SimulationContext():
+    yield SimulationContext_
+    SimulationContext_._clear_context_cache()
```
Comment (mattkappel) [inline tests/framework/test_engine.py:32]: Clever use of yield.

File: src/vivarium/framework/engine.py [hunk 2]
```diff
@@ -46,13 +46,39 @@
 
 
 class SimulationContext:
+
+    _created_simulation_contexts: Set[str] = set()
+
+    @staticmethod
+    def _get_context_name(sim_name: Union[str, None]) -> str:
+        # Not threadsafe, but I don't see why we'd need to be.
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:54]: You should add docstring making it clear that there is a side-effect here (i.e. adding the simulation to the set of simulation context names

File: tests/framework/test_engine.py (hunk 1 above)
Comment (rmudambi) [reply tests/framework/test_engine.py:32]: When does `SimulationContext_._clear_context_cache()` get triggered? I think it'd be helpful to add a comment explaining exactly what is happening here.

File: src/vivarium/framework/engine.py [hunk 3]
```diff
@@ -46,13 +46,39 @@
...
+            if sim_name in SimulationContext._created_simulation_contexts:
+                msg = (
+                    "Attempting to create two SimulationContexts "
+                    f"with the same name {sim_name}"
+                )
+                raise VivariumError(msg)
+
+        if sim_name is None:
+            sim_number = len(SimulationContext._created_simulation_contexts) + 1
+            sim_name = f"simulation_{sim_number}"
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:65]: What if there was a simulation manually named this? You should move this assignment above the error check.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium/framework/engine.py (hunk 3 above)
Comment (collijk) [reply src/vivarium/framework/engine.py:65]: good catch

File: tests/framework/test_engine.py (hunk 1 above)
Comment (collijk) [reply tests/framework/test_engine.py:32]: Disagree.  This is the standard teardown procedure for fixtures: https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended

**Review by rmudambi — APPROVED** [review]

**Comment by collijk** [comment]
Screwed up the commit history with an errant commit from a later PR.  See #287 for the redo

**Pull request closed without merging**